File: bindings/go/embedding.go
package latticedb

import cgobridge "github.com/jeffhajewski/latticedb/bindings/go/internal/cgo"

// Deprecated: use package github.com/jeffhajewski/latticedb/bindings/go/embedding.
type EmbeddingClient struct {
	raw *cgobridge.EmbeddingClient
}

// Deprecated: use package github.com/jeffhajewski/latticedb/bindings/go/embedding.
func HashEmbed(text string, dimensions uint16) ([]float32, error) {
	return cgobridge.HashEmbed(text, dimensions)
}

// Deprecated: use package github.com/jeffhajewski/latticedb/bindings/go/embedding.
func NewEmbeddingClient(config EmbeddingConfig) (*EmbeddingClient, error) {
	raw, err := cgobridge.NewEmbeddingClient(cgobridge.EmbeddingConfig{
		Endpoint:  config.Endpoint,
		Model:     config.Model,
		APIFormat: cgobridge.EmbeddingAPIFormat(config.APIFormat),
		APIKey:    config.APIKey,
		TimeoutMS: config.TimeoutMS,
	})
	if err != nil {
		return nil, wrapError(err)
	}
	return &EmbeddingClient{raw: raw}, nil
}

func (client *EmbeddingClient) Embed(text string) ([]float32, error) {
	if client == nil || client.raw == nil {
		return nil, ErrEmbeddingClosed
	}
	vector, err := client.raw.Embed(text)
	return vector, wrapError(err)
}

func (client *EmbeddingClient) Close() error {
	if client == nil || client.raw == nil {
		return nil
	}
	err := wrapError(client.raw.Close())
	if err == nil {
		client.raw = nil
	}
	return err
}

File: bindings/typescript/src/embedding.ts
/**
 * High-level embedding utilities for LatticeDB.
 *
 * Provides hash-based embeddings (built-in, no external service) and
 * an HTTP embedding client for services like Ollama and OpenAI.
 */

import { getFFI, EmbeddingClientHandle } from './ffi';
import { EmbeddingConfig } from './types';

export { EmbeddingApiFormat, EmbeddingConfig } from './types';

/**
 * Generate a hash embedding for text (built-in, no external service).
 *
 * Deterministic and fast. Useful for testing or simple keyword-based similarity.
 *
 * @param text - The text to embed.
 * @param dimensions - Number of embedding dimensions (default 128).
 * @returns A Float32Array of the specified dimensions.
 */
export function hashEmbed(text: string, dimensions: number = 128): Float32Array {
  return getFFI().hashEmbed(text, dimensions);
}

/**
 * HTTP embedding client for services like Ollama and OpenAI.
 *
 * @example
 * ```typescript
 * const client = new EmbeddingClient({ endpoint: "http://localhost:11434" });
 * const vec = client.embed("hello world");
 * client.close();
 * ```
 */
export class EmbeddingClient {
  private handle: EmbeddingClientHandle | null;

  /**
   * Create an embedding client connected to an HTTP service.
   *
   * @param config - Embedding service configuration.
   */
  constructor(config: EmbeddingConfig) {
    const ffi = getFFI();
    this.handle = ffi.embeddingClientCreate({
      endpoint: config.endpoint,
      model: config.model,
      apiFormat: config.apiFormat,
      apiKey: config.apiKey,
      timeoutMs: config.timeoutMs,
    });
  }

  /**
   * Generate an embedding for text via the HTTP service.
   *
   * @param text - The text to embed.
   * @returns A Float32Array embedding vector.
   */
  embed(text: string): Float32Array {
    if (this.handle === null) {
      throw new Error('EmbeddingClient has been closed');
    }
    return getFFI().embeddingClientEmbed(this.handle, text);
  }

  /**
   * Free the underlying native client. Safe to call multiple times.
   */
  close(): void {
    if (this.handle !== null) {
      getFFI().embeddingClientFree(this.handle);
      this.handle = null;
    }
  }
}
